### ISSUE TITLE
"virtual" is a mysql keyword but is used as a field in the tracks table

### ISSUE THREAD

**Issue by bkcabral** [LMS-Community/slimserver#197]
The use of "virtual" as a key word in newer mysql db means it can't be a field in a table. Unfortunately, it is used as such in the tracks table.

**Comment by mherger** [comment]
It's probably time to let go and just live with SQLite ;-). 

What's the reason to stick with MySQL for LMS?

**Comment by bkcabral** [comment]
The only arguments in the other direction is you want multiple music servers serving off a NAS + shared DB.  I don't have a need for that - and I assume most folks don't - but right now it's just broken.  So it should either be removed or fixed - no?

**Comment by bkcabral** [comment]
Didn't mean to close it... ;-)

**Comment by mherger** [comment]
That's correct: fix or remove. Would you happen to know in which version MySQL introduced the new keyword?

The fix might be as simple as search/replace virtual with something else (there aren't too many occurrences in our code), plus a schema migration. It's just not something I'd put high on my priority list...

**Comment by bkcabral** [comment]
I'm not sure when it was introduced but it's there in 5.7   FWIW - using the word "virtual" feels dangerous because it's too general and often "overloaded."  So it should probably be changed.  Now, I looked to see if was an "easy" search-and-replace and it's not - but I'm not sure how queries get formed - e.g. from hash key names in perl such that the scheme is implied in the hash keys.  Or I would just push a fix ;-)

**Comment by rossgrady** [comment]
Ha, I stopped by here to see how hard it would be to pull in multiple years worth of updates to my fork that I made specifically so I could continue to use MySQL . . . apologies to all involved for resurrecting this (still open) issue, but the fix here is as simple as backticking the word "virtual", like so:

Replace this:
```
			AND             virtual IS NULL
```
With this:
```
			AND             `virtual` IS NULL
```

**Comment by michaelherger** [comment]
Could you please confirm whether the following patch would fix this issue?

```patch
diff --git a/SQL/mysql/schema_12_down.sql b/SQL/mysql/schema_12_down.sql
index bcfb1cd36..453ee509b 100644
--- a/SQL/mysql/schema_12_down.sql
+++ b/SQL/mysql/schema_12_down.sql
@@ -1,7 +1,7 @@
 ALTER TABLE tracks DROP urlmd5;
 ALTER TABLE tracks DROP coverid;
 ALTER TABLE tracks DROP cover_cached;
-ALTER TABLE tracks DROP virtual;
+ALTER TABLE tracks DROP `virtual`;
 
 ALTER TABLE albums CHANGE artwork artwork int(10) default NULL;
 
diff --git a/SQL/mysql/schema_12_up.sql b/SQL/mysql/schema_12_up.sql
index a7a1c573b..e2bf014fa 100644
--- a/SQL/mysql/schema_12_up.sql
+++ b/SQL/mysql/schema_12_up.sql
@@ -15,7 +15,7 @@ CREATE INDEX coveridIndex ON tracks (coverid);
 
 ALTER TABLE tracks ADD cover_cached tinyint(1) default NULL;
 
-ALTER TABLE tracks ADD virtual tinyint(1) default NULL;
+ALTER TABLE tracks ADD `virtual` tinyint(1) default NULL;
 
 ALTER TABLE albums CHANGE artwork artwork char(8) default NULL;
 
diff --git a/Slim/Utils/Scanner/Local.pm b/Slim/Utils/Scanner/Local.pm
index 42665542c..8e8fa6feb 100644
--- a/Slim/Utils/Scanner/Local.pm
+++ b/Slim/Utils/Scanner/Local.pm
@@ -217,7 +217,7 @@ sub rescan {
 				WHERE filesize != 0
 			)
 			AND             url LIKE '$basedir%'
-			AND             virtual IS NULL
+			AND             `virtual` IS NULL
 			AND             content_type $ctFilter
 		} . (IS_SQLITE ? '' : ' ORDER BY url');
 
@@ -746,7 +746,7 @@ sub deleted {
 					SELECT id, album, year
 					FROM   tracks
 					WHERE  url LIKE '$url#%'
-					AND    virtual = 1
+					AND    `virtual` = 1
 				} );
 				$sth->execute;
 				$ptracks = $sth->fetchall_arrayref( {} );
```

(thanks @rossgrady!)

**Comment by rossgrady** [comment]
I can confirm that those are the only changes I needed to make on my fork of 7.9 in order to quash this problem. I made a few other MySQL-related changes because I was targeting a newer (at the time) version of MySQL but they were all Unicode-related.

FWIW, last night I cloned the latest 8.0 release & ran it straight vanilla, no mods, using SQLite, against my ~80,000-track library, and it was plenty performant, so I'm 99% sure I'm going to finally abandon MySQL with LMS! :)
